### PR TITLE
Fix cross validation for classification models and update pytorch-lightning

### LINKED ISSUE #433: No end-to-end AML job for classification model
Add a TrainClassificationCrossValidation training job in the PR build, similar to what is done for BasicModel2Epochs for segmentations

[AB#3952](https://innereye.visualstudio.com/60ce1777-00d6-4015-82bc-488a0c00202f/_workitems/edit/3952)

### PR DESCRIPTION
**Pull request by melanibe**
* Various fixes to metrics aggregation/download/paths for classification cross-validation runs:
** Fixed metrics aggregation for multi-gpu
** Fixed path for test metrics
** Added split recording for test set inference during CV
** Removed Ensemble from the folds taken into account for aggregated results in classification CV
** Removed Val inference for Ensemble (as val of parent fold is in training set of ensemble)
** Make CrossValidation folder follow a simple Train/ Val/ Test/ structure
* Add end-2-end test for classification model in PR build
* Updating pytorch-lightning to 1.2.7
* Fixed ddp setting when multi-node but only 1 gpu per node
* WARNING: removing test for training multi-node as new PL version breaks multi-node training.
Closes #433

### CODE REVIEW

**Review by ant0nsc — APPROVED** [review]

File: InnerEye/ML/deep_learning_config.py [hunk 1]
```diff
@@ -295,9 +295,7 @@ class DeepLearningConfig(GenericConfig, CudaAwareConfig):
                          "to that set.",
                      allow_None=True)
     perform_training_set_inference: bool = \
-        param.Boolean(False,
-                      doc="If False (default), run full image inference on validation and test set after training. If "
-                          "True, also run full image inference on the training set")
+        param.Boolean(False, doc="If True, run full image inference on the training set at the end of training.")
```
[the remark this answers is not in the document]
Comment (melanibe) [reply InnerEye/ML/deep_learning_config.py:298]: Re-extend. I did not like the previous version because it implied that this flag=False means inference on val/test automatically but in reality it is controlled by the next flag.

File: InnerEye/ML/metrics_dict.py [hunk 2]
```diff
@@ -490,7 +491,7 @@ def get_cross_entropy(self, hue: str = DEFAULT_HUE_KEY) -> float:
         """
         predictions = self.get_predictions(hue)
         labels = self.get_labels(hue)
-        return log_loss(labels, predictions)
+        return log_loss(labels, predictions, labels=[0, 1])
```
Comment (melanibe) [inline InnerEye/ML/metrics_dict.py:494]: Needed in case we only see one label (e.g. in the tests)

File: InnerEye/ML/deep_learning_config.py [hunk 3]
```diff
@@ -381,10 +382,10 @@ class DeepLearningConfig(GenericConfig, CudaAwareConfig):
...
     #: Name of the csv file providing information on the dataset to be used.
     dataset_csv: str = param.String(
-            DATASET_CSV_FILE_NAME,
-            doc="Name of the csv file providing information on the dataset "
-                "to be used, containing at least the fields: "
-                "subject, channel, filePath.")
+        DATASET_CSV_FILE_NAME,
+        doc="Name of the csv file providing information on the dataset "
+            "to be used, containing at least the fields: "
+            "subject, channel, filePath.")
```
Comment (ant0nsc) [inline InnerEye/ML/deep_learning_config.py:388]:
```suggestion
        doc="Name of the CSV file providing information on the dataset to be used. "
            "For segmentation models, this file must contain at least the fields: `subject`, `channel`, `filePath`.")
```

File: InnerEye/ML/metrics_dict.py [hunk 4]
```diff
@@ -662,14 +664,17 @@ def store_metrics_per_subject(self,
...
+                    LoggingColumns.Epoch.value: epoch,
                     LoggingColumns.CrossValidationSplitIndex.value: cross_validation_split_index,
                     LoggingColumns.DataSplit.value: mode.value
                 })
 
     @staticmethod
     def load_execution_mode_metrics_from_df(df: pd.DataFrame,
                                             is_classification_metrics: bool) -> Dict[ModelExecutionMode,
-                                                                                     Dict[int, ScalarMetricsDict]]:
+                                                                                     Dict[Union[
```
Comment (ant0nsc) [inline InnerEye/ML/metrics_dict.py:675]: you are using the very complex thing repeatedly, define a type alias?

File: InnerEye/ML/model_training.py [hunk 5]
```diff
@@ -290,3 +280,24 @@ def model_train(config: ModelConfigBase,
         resource_monitor.kill()
 
     return model_training_results
+
+
+def aggregate_and_create_subject_metrics_file(config: ModelConfigBase) -> None:
+    """
+    This functions takes all the subject metrics file written by each GPU (one file per GPU) and aggregates them into
+    one single metrics file. Results is saved in config.outputs_folder / mode.value / SUBJECT_METRICS_FILE_NAME
```
Comment (ant0nsc) [inline InnerEye/ML/model_training.py:288]:
```suggestion
    This functions takes all the subject metrics file written by each GPU (one file per GPU) and aggregates them into
    one single metrics file. Results is saved in config.outputs_folder / mode.value / SUBJECT_METRICS_FILE_NAME.
    This is done for the metrics files for training and for validation data separately.
```

File: InnerEye/ML/model_training.py [hunk 6]
```diff
@@ -290,3 +280,24 @@ def model_train(config: ModelConfigBase,
         resource_monitor.kill()
 
     return model_training_results
+
+
+def aggregate_and_create_subject_metrics_file(config: ModelConfigBase) -> None:
```
Comment (ant0nsc) [inline InnerEye/ML/model_training.py:285]: The signature could be simplified (and hence testing made easier) if it only consumed an outputs_folder argument.

File: azure-pipelines/build-pr.yml [hunk 7]
```diff
@@ -124,3 +124,27 @@ jobs:
         parameters:
           pytest_mark: after_training_2node
           test_run_title: tests_after_training_2node_run
+
+  # Train a model on 2 nodes
```
Comment (ant0nsc) [inline azure-pipelines/build-pr.yml:128]:
```suggestion
  # Train a classification model in cross validation mode
```

**Review by ant0nsc — CHANGES_REQUESTED** [review]
Good stuff, thanks Melanie! just a few small niggles.

File: InnerEye/ML/metrics_dict.py (hunk 4 above)
Comment (melanibe) [reply InnerEye/ML/metrics_dict.py:675]: updated

File: InnerEye/ML/metrics_dict.py [hunk 8]
```diff
@@ -824,3 +827,6 @@ def flush(self, log_info: bool = False) -> None:
         if log_info:
             s = df.to_string(index=False, float_format="%.6f")
             logging.info(f"\n{s}")
+
+
+MetricsPerExecutionModeAndEpoch = Dict[ModelExecutionMode, Dict[Union[int, str], ScalarMetricsDict]]
```
Comment (dccastro) [inline InnerEye/ML/metrics_dict.py:832]: perhaps better defined at the top of the file?

File: InnerEye/ML/visualizers/plot_cross_validation.py [hunk 9]
```diff
@@ -36,14 +36,14 @@
 from InnerEye.Common.Statistics.wilcoxon_signed_rank_test import WilcoxonTestConfig, wilcoxon_signed_rank_test
 from InnerEye.Common.common_util import CROSSVAL_RESULTS_FOLDER, ENSEMBLE_SPLIT_NAME, \
     FULL_METRICS_DATAFRAME_FILE, \
-    METRICS_AGGREGATES_FILE, OTHER_RUNS_SUBDIR_NAME, logging_section, logging_to_stdout
+    METRICS_AGGREGATES_FILE, ModelProcessing, logging_section, logging_to_stdout
```
Comment (dccastro) [inline InnerEye/ML/visualizers/plot_cross_validation.py:39]: reformat?

File: environment.yml [hunk 10]
```diff
@@ -43,7 +44,7 @@ dependencies:
       - pytest-cov==2.10.1
       - pytest-forked==1.3.0
       - pytest-xdist==1.34.0
-      - pytorch-lightning==1.1.8
+      - pytorch-lightning==1.2.4
```
Comment (dccastro) [inline environment.yml:47]: was it not 1.2.7?

File: Tests/ML/models/test_scalar_model.py [hunk 11]
```diff
@@ -14,8 +14,9 @@
 import torch
 
 from InnerEye.Common import common_util, fixed_paths
-from InnerEye.Common.common_util import CROSSVAL_RESULTS_FOLDER, EPOCH_METRICS_FILE_NAME, METRICS_AGGREGATES_FILE, \
-    SUBJECT_METRICS_FILE_NAME, get_epoch_results_path, logging_to_stdout
+from InnerEye.Common.common_util import BEST_EPOCH_FOLDER_NAME, CROSSVAL_RESULTS_FOLDER, EPOCH_METRICS_FILE_NAME, \
+    METRICS_AGGREGATES_FILE, \
+    SUBJECT_METRICS_FILE_NAME, get_best_epoch_results_path, logging_to_stdout
```
Comment (dccastro) [inline Tests/ML/models/test_scalar_model.py:19]: reformat?

File: Tests/ML/visualizers/test_plot_cross_validation.py [hunk 12]
```diff
@@ -11,7 +11,8 @@
 from pandas.core.dtypes.common import is_string_dtype
 
 from InnerEye.Azure.azure_util import CROSS_VALIDATION_SPLIT_INDEX_TAG_KEY
-from InnerEye.Common.common_util import CROSSVAL_RESULTS_FOLDER, FULL_METRICS_DATAFRAME_FILE, METRICS_AGGREGATES_FILE, \
+from InnerEye.Common.common_util import CROSSVAL_RESULTS_FOLDER, FULL_METRICS_DATAFRAME_FILE, \
+    METRICS_AGGREGATES_FILE, \
     SUBJECT_METRICS_FILE_NAME, logging_to_stdout
```
Comment (dccastro) [inline Tests/ML/visualizers/test_plot_cross_validation.py:16]: reformat?

File: environment.yml (hunk 10 above)
Comment (melanibe) [reply environment.yml:47]: Yeah both work - i can go 1.2.7.

File: InnerEye/ML/metrics_dict.py (hunk 8 above)
Comment (melanibe) [reply InnerEye/ML/metrics_dict.py:832]: nope cause we need ScalarMetricsDict to be defined first

File: InnerEye/ML/model_training.py [hunk 13]
```diff
@@ -290,3 +280,25 @@ def model_train(config: ModelConfigBase,
...
+        result_file = outputs_folder / mode.value / SUBJECT_METRICS_FILE_NAME
+        result_file = result_file.open("a")
+        for i, file in enumerate(temp_files):
+            temp_file_contents = file.read_text()
+            if i == 0:
+                # Copy the first file as-is, including the first line with the column headers
+                result_file.write(temp_file_contents)
+            else:
+                # For all files but the first one, cut off the header line.
+                result_file.write("\n" + os.linesep.join(temp_file_contents.splitlines()[1:]))
```
Comment (dccastro) [inline InnerEye/ML/model_training.py:303]: inconsistent use of `'\n'` and `os.linesep`

Comment (melanibe) [reply InnerEye/ML/model_training.py:303]: good catch!

File: InnerEye/ML/model_testing.py [hunk 14]
```diff
@@ -13,9 +13,10 @@
 import matplotlib.pyplot as plt
 import numpy as np
 
-from InnerEye.Azure.azure_util import PARENT_RUN_CONTEXT
-from InnerEye.Common.common_util import METRICS_AGGREGATES_FILE, ModelProcessing, SUBJECT_METRICS_FILE_NAME, \
-    get_epoch_results_path, is_linux, logging_section
+from InnerEye.Azure.azure_util import DEFAULT_CROSS_VALIDATION_SPLIT_INDEX, PARENT_RUN_CONTEXT
+from InnerEye.Common.common_util import BEST_EPOCH_FOLDER_NAME, METRICS_AGGREGATES_FILE, ModelProcessing, \
+    SUBJECT_METRICS_FILE_NAME, \
+    get_best_epoch_results_path, is_linux, logging_section
```
Comment (dccastro) [inline InnerEye/ML/model_testing.py:19]: reformat?

**Review by dccastro — COMMENTED** [review]
Looks great to me! Just proposed very superficial changes.

File: InnerEye/ML/metrics_dict.py (hunk 8 above)
Comment (dccastro) [reply InnerEye/ML/metrics_dict.py:832]: but you can use a forward type declaration as a string: `'ScalarMetricsDict'`

Comment (melanibe) [reply InnerEye/ML/metrics_dict.py]: Updated :)

**Review by dccastro — DISMISSED** [review]
Awesome, all clear 👍

**Review by ant0nsc — DISMISSED** [review]

**Review by ant0nsc — APPROVED** [review]

**Review by dccastro — APPROVED** [review]
